File: packages/rayact-haptics/android/src/main/java/dev/rayact/haptics/RayactHapticsRegistration.kt
package dev.rayact.haptics

import android.view.HapticFeedbackConstants
import com.rayact.engine.RayactPlatformModuleRegistration
import com.rayact.engine.RayactPlatformRegistry
import org.json.JSONObject

class RayactHapticsRegistration : RayactPlatformModuleRegistration {
    override fun register(
        context: android.content.Context,
        registry: RayactPlatformRegistry,
    ) {
        registry.registerModule("haptics") { method, payload, completion ->
            runCatching {
                val activity = RayactPlatformRegistry.currentActivity()
                    ?: error("No foreground activity is available")
                val data = runCatching { JSONObject(payload.toString(Charsets.UTF_8)) }
                    .getOrElse { JSONObject() }
                val feedback = when (method) {
                    "selection" -> HapticFeedbackConstants.TEXT_HANDLE_MOVE
                    "impact" -> when (data.optString("style")) {
                        "heavy", "rigid" -> HapticFeedbackConstants.LONG_PRESS
                        else -> HapticFeedbackConstants.CONTEXT_CLICK
                    }
                    "notification" -> when (data.optString("type")) {
                        "error" -> HapticFeedbackConstants.REJECT
                        else -> HapticFeedbackConstants.CONFIRM
                    }
                    else -> error("Unknown haptics method '$method'")
                }
                activity.runOnUiThread {
                    activity.window.decorView.performHapticFeedback(feedback)
                }
                "true".toByteArray(Charsets.UTF_8)
            }.fold(
                onSuccess = { completion(Result.success(it)) },
                onFailure = { completion(Result.failure(it)) },
            )
        }
    }
}

File: packages/rayact-haptics/ios/RayactHapticsRegistration.swift
import Foundation
import UIKit

final class RayactHapticsRegistration: RayactPlatformModuleRegistration {
    func register(with registry: RayactPlatformRegistry) {
        registry.registerModule("haptics") { method, payload, completion in
            let value = (try? JSONSerialization.jsonObject(with: payload)) as? [String: Any] ?? [:]
            let perform = {
                switch method {
                case "selection":
                    UISelectionFeedbackGenerator().selectionChanged()
                case "impact":
                    let style: UIImpactFeedbackGenerator.FeedbackStyle
                    switch value["style"] as? String {
                    case "light": style = .light
                    case "heavy": style = .heavy
                    case "soft": style = .soft
                    case "rigid": style = .rigid
                    default: style = .medium
                    }
                    UIImpactFeedbackGenerator(style: style).impactOccurred()
                case "notification":
                    let type: UINotificationFeedbackGenerator.FeedbackType
                    switch value["type"] as? String {
                    case "warning": type = .warning
                    case "error": type = .error
                    default: type = .success
                    }
                    UINotificationFeedbackGenerator().notificationOccurred(type)
                default:
                    completion(.failure(NSError(
                        domain: "RayactHaptics",
                        code: 1,
                        userInfo: [NSLocalizedDescriptionKey: "Unknown haptics method '\(method)'"]
                    )))
                    return
                }
                completion(.success(Data("true".utf8)))
            }
            if Thread.isMainThread {
                perform()
            } else {
                DispatchQueue.main.async(execute: perform)
            }
        }
    }
}
